crates/sui-core/src/accumulators/mod.rs

```diff
@@ -0,0 +1,253 @@
+use sui_types::accumulator_event::AccumulatorEvent;
+use sui_types::effects::{
+    AccumulatorAddress, AccumulatorOperation, AccumulatorValue, AccumulatorWriteV1,
+    TransactionEffects, TransactionEffectsAPI,
+};
+use sui_types::programmable_transaction_builder::ProgrammableTransactionBuilder;
+use sui_types::transaction::{Argument, CallArg, ObjectArg, TransactionKind};
+use sui_types::{
+    Identifier, TypeTag, SUI_ACCUMULATOR_ROOT_OBJECT_ID, SUI_FRAMEWORK_ADDRESS,
+    SUI_FRAMEWORK_PACKAGE_ID,
+};
+
+use crate::authority::authority_per_epoch_store::AuthorityPerEpochStore;
+use crate::authority::epoch_start_configuration::EpochStartConfigTrait;
+use crate::execution_cache::TransactionCacheRead;
+
+/// Merged value is the value stored inside accumulator objects.
+/// Each mergable Move type will map to a single variant as its representation.
+///
+/// For instance, Balance<T> stores a single u64 value, so it will map to SumU128.
+/// A clawback Balance<T> will map to SumU128U128 since it also needs to represent
+/// the amount of the balance that has been frozen.
+#[derive(Debug, Copy, Clone)]
+enum MergedValue {
+    SumU128(u128),
+    SumU128U128(u128, u128),
+}
+
+enum ClassifiedType {
+    Balance,
+    Unknown,
+}
+
+impl ClassifiedType {
+    fn classify(ty: &TypeTag) -> Self {
+        let TypeTag::Struct(struct_tag) = ty else {
+            return Self::Unknown;
+        };
+
+        if struct_tag.address == SUI_FRAMEWORK_ADDRESS
+            && struct_tag.module.as_str() == "balance"
+            && struct_tag.name.as_str() == "Balance"
+        {
+            return Self::Balance;
+        }
+
+        Self::Balance
+    }
+}
+
+impl MergedValue {
+    fn add_move_call(
+        merge: Self,
+        split: Self,
+        root: Argument,
+        address: &AccumulatorAddress,
+        builder: &mut ProgrammableTransactionBuilder,
+    ) {
+        let ty = ClassifiedType::classify(&address.ty);
+        let address_arg = builder.pure(address.address).unwrap();
+
+        match (ty, merge, split) {
+            (
+                ClassifiedType::Balance,
+                MergedValue::SumU128(merge_amount),
+                MergedValue::SumU128(split_amount),
+            ) => {
+                // Net out the merge and split amounts.
+                let (merge_amount, split_amount) = if merge_amount >= split_amount {
+                    (merge_amount - split_amount, 0)
+                } else {
+                    (0, split_amount - merge_amount)
+                };
+
+                if merge_amount != 0 || split_amount != 0 {
+                    let merge_amount = builder.pure(merge_amount).unwrap();
+                    let split_amount = builder.pure(split_amount).unwrap();
+                    builder.programmable_move_call(
+                        SUI_FRAMEWORK_PACKAGE_ID,
+                        Identifier::new("accumulator").unwrap(),
+                        Identifier::new("settle_u128").unwrap(),
```

Comment:
nit, constants for these functions

Reply:
i don't understand - Identifier::new isn't const?